### DISCUSSION TITLE
Qt Creator terminal integration

### DISCUSSION THREAD

**Discussion by cristianadam** [chrisant996/clink#453] [Q&A] [answered]
Qt Creator 11 will get  terminal integration. I would like to have the best `clink` support out of the box.

Currently there are two issues:
1. The light theming is not working nicely with clink
2. The shell integration (mostly getting current directory) is missing

See the screencast below:
![qtcreator-clink-light-theme](https://user-images.githubusercontent.com/1257414/231450239-078d7414-e24a-4539-af13-1b814c3a3f90.gif)

The shell integration we took (MIT licensed) from VS Code https://github.com/microsoft/vscode/tree/main/src/vs/workbench/contrib/terminal/browser/media

How should Qt Creator change the environment where `cmd.exe` is launched? So that the user wouldn't have to change the clink configuration by him/herself.

**Comment by chrisant996** [comment]
Re: "light theme" -- see the (long) discussion in #256.

Re: "shell integration so the user doesn't have to configure Clink" -- I think you mean you want a _**default**_ "shell integration" in case the user hasn't configured Clink at all yet. I don't think you should try to override/replace user configuration.

Can you describe what kinds of shell integration are desired?

If you just want to set a default prompt string, that's not a Clink question, it's a CMD question. Run `prompt /?` for some info.

**Reply by cristianadam** [reply]
Oh, #256 is a bit scary . We'll have to see how we can fix it from our side. Documenting things would also be fine I think.

The shell integration means that the shell executable will have to load some scripts (bash, ps1) and send OSC (Operating System Command) codes back to the terminal application.

In the screencast above you can see that bash has the title "bash - \<current working directory\>".

Something in `lua` would be need for `clink` to have the same functionality.

**Reply by chrisant996** [reply]
> Oh, #256 is a bit scary . We'll have to see how we can fix it from our side. Documenting things would also be fine I think.
> 
> The shell integration means that the shell executable will have to load some scripts (bash, ps1) and send OSC (Operating System Command) codes back to the terminal application.
> 
> In the screencast above you can see that bash has the title "bash - <current working directory>".
> 
> Something in `lua` would be need for `clink` to have the same functionality.

I meant what are the specific shell integration behaviors that are needed.  I spent an hour reading the scripts to learn the answers.

The environment variable "replace/prepend/append" stuff is better done in a .cmd shell script, not in a Lua script.

The custom history file is easy; simply set [%CLINK_HISTORY_LABEL%](https://chrisant996.github.io/clink/clink.html#multiple-master-history-files).

The "shell integration" is primarily about printing escape codes at specific points:
- send escape code before prompt, after prompt
- send escape code before R-prompt, after R-prompt
- send escape code before executing a command line
- send escape codes after executing a command line, telling the host what the command's exit code was.

Those indeed require a Lua script.  But Clink allows prompt filters to halt the filtering, which means currently there is not a reliable way to hook the prompt to add escape codes before and after the prompt.  And Clink doesn't have a way to say "start Clink with the user's normal configuration, BUT ALSO include the following script(s) or script directory(ies)".

So it would be necessary to add a way to either let a prompt filter supersede the "halt filtering" directive, or to add an event to allow post-processing the prompt strings.  And a way to add an extra script to run when injecting Clink.

I'll think more about those and reply again in a few days when I have a more clear sense of how that would work best.

**Reply by chrisant996** [reply]
@cristianadam The following questions matter, because they require different new things from Clink:

I haven't noticed any problems in VSCode when using Clink, so I'm a little unclear on why escape codes are needed before/after the prompt.  What does the terminal host intend to do in response to the escape codes?

**Is it about timing?**
- If Qt Creator wants to know "the prompt is about to be displayed" and "input line editing is about to begin"
- Then Clink needs a new event for _after_ the prompt has been displayed.
- But also that would be a deficiency in the terminal host; the host cannot safely assume that the prompt is printed only once.

**Is it about position?**
- If Qt Creator wants to know literally "the left prompt begins here" and "the left prompt ends here" and "the right prompt begins here" and "the right prompt ends here"
- Then one of the following is needed:
  - Clink needs a way to specify escape codes as prefix/suffix around the left and right prompt strings.  This is fragile because it only lets ONE thing choose what escape codes should be used.  (And it should probably strictly enforce that _only_ escape codes are allowed; no text.)
  - Or Clink needs a way for a prompt filter to request to be called even if prompt filtering is halted early.  This is fragile because another prompt filter could still run after the one that tries to add prefix/suffix escape codes, and then they won't actually be proper prefix or suffix.
  - Or Clink needs a new event for _before_ the prompt is displayed, to enable modifying the prompt strings.  This is fragile because another event handler could still run after the one that tries to add prefix/suffix escape codes.  Or if it allows an event handler to cancel subsequent event handlers, then it's fragile because the one for Qt Creator might not even be reached.

**Reply by chrisant996** [reply]
@cristianadam I've added an optional `:surround()` function that prompt filters can define.  Then Qt Creator could define a prompt filter that doesn't actually modify the prompt text, but surrounds the prompt text with prefix/suffix escape codes.  It can use a negative priority (e.g. `clink.promptfilter(-999)`) to ensure that it runs before other prompt filters, so that it takes precedence over whatever other prompt filters do.

Next I'll work on a way to add an extra Lua script when injecting Clink.

UPDATE:  I'm also adding environment variables for defining prompt prefix/suffix strings.  That would almost completely eliminate the need for a Lua script.  Hopefully I can find some way to enable expanding the CWD inside prefix/suffix strings -- if so, then there wouldn't need to be any Lua script at all.  The shell integration could be done simply with a CMD shell script (.cmd file), which would set %CLINK_PROMPT_PREFIX% / etc environment variables.

**Reply by chrisant996** [reply]
@cristianadam The `%CLINK_PROMPT_PREFIX%` / etc strings now also expand CMD prompt `$` codes.  So, embedding `$p` in a prompt prefix or suffix string expands to the current working directory.  So, no Lua needed for the prompt prefix/suffix strings or even for setting the current working directory into the terminal title.

But it looks like Qt Creator (and VSCode) also want to receive escape codes when command line execution starts and finishes.  Currently that would require a Lua script.  I'll investigate what to do about that.  Maybe the best thing will be to add support for injecting an extra Lua script when injecting Clink, or maybe another mechanism will be possible.

**Reply by cristianadam** [reply]
The Qt Creator shell integration part is done at https://github.com/qt-creator/qt-creator/blob/master/src/plugins/terminal/shellintegration.cpp#L70 and it covers the current directory / current command line.

I should have mentioned that my `cmd.exe` has a custom prompt (`PROMPT=$p_$$$s`), but that's independent from the task at hand. I can see how this can cause confusion.

**Reply by chrisant996** [reply]
@cristianadam The best way to load a "shell integration" Lua script is by setting (appending to) the %CLINK_PATH% environment variable.
- Using `inject --scripts` would break a Cmder shell instance inside the editor's terminal window.
- Using %CLINK_PATH% will be compatible with existing shell instances, _unless_ the user (or the shell configuration) explicitly replaces %CLINK_PATH% instead of appending to it.

Interestingly, it looks like with Clink v1.4.24 and higher (which will be published in a few days), almost all of the "shell integration" can be done without a Lua script.  The only thing that seems to really require a Lua script is sending the exit code.

EXAMPLE SCRIPTS

**clink_integration.cmd**
```cmd
set CLINK_HISTORY_LABEL=QtCreator
set CLINK_PATH=%CLINK_PATH%;c:\qtcreator_path\script\clink_integration.lua
rem -- ADD CODE HERE to do any desired environment variable replacements.
```

**clink_integration.lua**
```lua
local p = clink.promptfilter(-999)

function p:filter()
    -- nothing to do here, but it needs to be defined.
end

function p:surround()
    local pre, suf
    local rpre, rsuf

    -- esc codes surrounding prompt string
    pre = "\027]633;A\a" -- copied from shellIntegration-rc.zsh
    suf = "\027]633;B\a" -- copied from shellIntegration-rc.zsh
    suf = suf .. "\027]633;D;" .. os.geterrorlevel() .. "\a" -- send command exit code
    suf = suf .. "\027]633;P;Cwd=" .. os.getcwd() .. "\a" -- send cwd as title

    -- esc codes surrounding right side prompt string
    rpre = "\027]633;H\a" -- copied from shellIntegration-rc.zsh
    rsuf = "\027]633;I\a" -- copied from shellIntegration-rc.zsh

    return pre, suf, rpre, rsuf
end
```

But again, when I read the vscode shell integration scripts, I'm a little concerned:  it appears as though the scripts might be assuming that a shell prints the prompt exactly once per command line.  If so, that would be an inaccurate assumption:  bash, fish, zsh, and clink in cmd all can reprint the prompt multiple times per command line, e.g. after showing completions or in response to background work finishing and needing to update the prompt text.  So, it matters how the terminal host (e.g. Qt Creator) expects to respond to the escape codes.  I don't have time available to analyze Qt Creator code to figure that out, so my ability to give specific guidance is limited.

**Reply by cristianadam** [reply]
Qt Creator just reacts on the OSC codes when they come. I think the timing is not important. 

Also the console output is not affected by this information. When the OSC codes are received then the appropriate UI elements would be updated.

**Reply by chrisant996** [reply]
> The Qt Creator shell integration part is done at https://github.com/qt-creator/qt-creator/blob/master/src/plugins/terminal/shellintegration.cpp#L70 and it covers the current directory / current command line.

Ah, interesting.  Most of the escape codes from the vscode shell integration scripts are ignored (e.g. `ESC]633;` followed by `A`, `B`, `F`, `G`, `H`, or `I`) -- those are the codes where the "timing vs position" question is relevant, but none of them are handled by Qt Creator, so the question ends up moot. 😎

That simplifies things.  So it's really just the `D` and `P;Cwd=` codes.  See the example scripts I shared in my previous reply.  And the other codes could even be removed from the scripts.

**Reply by cristianadam** [reply]
I'll try it out next days and provide feedback. Thank you for the support!

**Reply by chrisant996** [reply, marked as answer]
@cristianadam Oh!  It's even easier!  Since Qt Creator only wants exit code and cwd, that's already fully possible with Clink as-is, for probably at least a year.

Instead of using a prompt filter, just use `onbeginedit`:

```lua
local function send_codes()
    local codes = ""
    codes = codes .. "\027]633;D;" .. os.geterrorlevel() .. "\a" -- send command exit code
    codes = codes .. "\027]633;P;Cwd=" .. os.getcwd() .. "\a" -- send cwd as title
    clink.print(codes, NONL)
end

clink.onbeginedit(send_codes)
```

And use the example .cmd script from the [earlier reply](https://github.com/chrisant996/clink/discussions/453#discussioncomment-5640049) to load the Lua script.

**Comment by chrisant996** [comment]
I've made a script for vscode shell integration, which I'll publish in the clink-gizmos repo soon.

Here it is.  Presumably Qt Creator will set `%TERM_PROGRAM%` to something else, so you could model a script after this, checking for something other than `"vscode"`.

**Important:** Be sure to rename the `gizmo_vscode_shell_integration` variable, so that your script doesn't collide with this one.

```lua
-- luacheck: globals gizmo_vscode_shell_integration NONL
if gizmo_vscode_shell_integration == nil then
    gizmo_vscode_shell_integration = true
end

if not gizmo_vscode_shell_integration then
    return
end

local function is_vscode()
    local term_program = os.getenv("term_program") or ""
    if term_program:lower() == "vscode" then
        return true
    end
end

local function send_context()
    if is_vscode() then
        local codes = ""
        codes = codes .. "\027]633;D;" .. os.geterrorlevel() .. "\a" -- send command exit code
        codes = codes .. "\027]633;P;Cwd=" .. os.getcwd() .. "\a" -- send cwd as title
        clink.print(codes, NONL)
    end
end

local p = clink.promptfilter(-999)

function p:filter() -- luacheck: no unused
    -- Nothing to do here, but the filter function must be defined.
end

function p:surround() -- luacheck: no unused
    if is_vscode() then
        local pre, suf
        local rpre, rsuf

        -- ESC codes surrounding prompt string
        pre = "\027]633;A\a" -- copied from shellIntegration-rc.zsh
        suf = "\027]633;B\a" -- copied from shellIntegration-rc.zsh

        -- ESC codes surrounding right side prompt string
        rpre = "\027]633;H\a" -- copied from shellIntegration-rc.zsh
        rsuf = "\027]633;I\a" -- copied from shellIntegration-rc.zsh

        return pre, suf, rpre, rsuf
    end
end

clink.onbeginedit(send_context)
```

**Comment by cristianadam** [comment]
![qtcreator-terminal-clink-shellintegration](https://user-images.githubusercontent.com/1257414/233161734-dc7caf8a-ecf6-4dc4-8a98-0ee0de139780.gif)

**Reply by chrisant996** [reply]
Glad you got it working.
